src/ucs/datastruct/array.inl

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_ARRAY_INL_
+#define UCS_ARRAY_INL_
+
+#include <ucs/datastruct/array.h>
+#include <ucs/sys/math.h>
+#include <ucs/debug/log.h>
+#include <ucs/debug/memtrack.h>
+
+
+/* Increase the array buffer length by this factor, whenever it needs to grow */
+#define UCS_ARRAY_GROW_FACTOR   2
+
+
+/**
+ * Define the functions of an array
+ *
+ * @param _name        Array name
+ * @param _index_type  Type of array's index
+ * @param _value_type  Type of array's values
+ * @param _scope       Scope for array's functions (e.g 'static inline')
+ */
+#define UCS_ARRAY_IMPL(_name, _index_type, _value_type, _scope) \
+    \
+    _scope void \
+    UCS_ARRAY_IDENTIFIER(_name, _init_dynamic)(ucs_array_t(_name) *array) \
+    { \
+        array->buffer   = NULL; \
+        array->length   = 0; \
+        array->capacity = 0; \
+    } \
+    \
+    _scope void \
+    UCS_ARRAY_IDENTIFIER(_name, _cleanup_dynamic)(ucs_array_t(_name) *array) \
+    { \
+        ucs_assert(!ucs_array_is_fixed(array)); \
+        ucs_free(array->buffer); \
+    } \
+    \
+    _scope ucs_status_t \
+    UCS_ARRAY_IDENTIFIER(_name, _grow)(ucs_array_t(_name) *array, \
+                                       _index_type min_capacity) \
+    { \
+        _index_type new_capacity; \
+        _value_type *new_buffer; \
+        size_t alloc_length; \
+        \
+        if (ucs_array_is_fixed(array)) { \
+            ucs_debug("cannot grow fixed array %s from %zu to %zu elems", \
+                      UCS_PP_MAKE_STRING(_name), (size_t)ucs_array_capacity(array), \
+                      (size_t)min_capacity); \
+            return UCS_ERR_NO_MEMORY; \
+        } \
+        \
+        new_capacity = ucs_max(array->capacity * UCS_ARRAY_GROW_FACTOR, \
+                               min_capacity); \
+        new_capacity = (new_capacity + ~UCS_ARRAY_CAP_MASK) & UCS_ARRAY_CAP_MASK; \
+        \
+        alloc_length = sizeof(_value_type) * new_capacity; \
+        new_buffer   = (_value_type*)ucs_realloc(array->buffer, alloc_length, \
+                                                 UCS_PP_MAKE_STRING(_name)); \
+        if (new_buffer == NULL) { \
+            ucs_debug("failed to grow %s from %zu to %zu elems of '%s'", \
```

Comment:
maybe use `error` scope?

Reply:
not sure, e.g for string buffer case it can check for error status

Reply:
yes, but should not we treat memory allocation problem as an error? We can handle return code properly, but still print warn/error

Reply:
ok